src/librustc_typeck/collect.rs

```diff
@@ -1085,12 +1085,16 @@ fn type_of<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
 
         NodeField(field) => icx.to_ty(&field.ty),
 
-        NodeExpr(&hir::Expr { node: hir::ExprClosure(.., gen), .. }) => {
+        NodeExpr(&hir::Expr { node: hir::ExprClosure(.., gen), span, .. }) => {
             if gen.is_some() {
                 let hir_id = tcx.hir.node_to_hir_id(node_id);
                 return tcx.typeck_tables_of(def_id).node_id_to_type(hir_id);
             }
 
+            if !tcx.has_typeck_tables(def_id) {
```

Comment:
Do we still need this check? I thought the DefIds were setup correctly now?

Reply:
The def ids in discriminants will still have the enum as parent (so will fail here), for structs a case like `struct Foo([u8; |x: u8| { }])` will also fail because the closure itself is the outer expression so it will have no parent to attach. So yhea without the check this would still ICE in some cases.

Reply:
oh... I thought the changes that made `(|x: u8| {}, 42).1` work also got rid of the ICE.